File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/A-Суммы_и_XOR-ы_на_отрезках/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    prefix_sum: list[int] = [0] * (n + 1)
    prefix_xor: list[int] = [0] * (n + 1)
    nums: iter[int] = map(int, sys.stdin.readline().rstrip().split())
    for i in range(n):
        num: int = next(nums)
        prefix_sum[i + 1] = prefix_sum[i] + num
        prefix_xor[i + 1] = prefix_xor[i] ^ num
    m: int = int(sys.stdin.readline().rstrip())
    for _ in range(m):
        op, l, r = map(int, sys.stdin.readline().rstrip().split())
        if op == 1: sys.stdout.write('{}\n'.format(prefix_sum[r] - prefix_sum[l - 1]))
        else: sys.stdout.write('{}\n'.format(prefix_xor[r] ^ prefix_xor[l - 1]))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/A-Суммы_и_XOR-ы_на_отрезках/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<long long> prefixSum (n + 1, 0);
    std::vector<long long> prefixXOR (n + 1, 0);
    for (int i = 0; i < n; ++i) {
        long long num;
        std::scanf("%lld", &num);
        prefixSum[i + 1] = prefixSum[i] + num;
        prefixXOR[i + 1] = prefixXOR[i] ^ num;
    }
    int m;
    std::scanf("%d", &m);
    for (int i = 0; i < m; ++i) {
        int op, l, r;
        std::scanf("%d %d %d", &op, &l, &r);
        if (op == 1) std::printf("%lld\n", prefixSum[r] - prefixSum[l - 1]);
        else std::printf("%lld\n", prefixXOR[r] ^ prefixXOR[l - 1]);
    }
}


int main() {
    solution();
}